pandas/_libs/lib.pyx

```diff
@@ -2227,123 +2371,146 @@ def maybe_convert_numeric(
         allow_null_in_int = convert_to_masked_nullable and not seen.float_
 
         if val.__hash__ is not None and val in na_values:
-            if allow_null_in_int:
-                seen.null_ = True
+            if allow_null_in_int or convert_to_masked_nullable:
                 mask[i] = 1
-            else:
-                if convert_to_masked_nullable:
-                    mask[i] = 1
-                seen.saw_null()
-            floats[i] = complexes[i] = NaN
+            seen.saw_null(seen_float=not allow_null_in_int)
         elif util.is_float_object(val):
             fval = val
             if fval != fval:
-                seen.null_ = True
                 if allow_null_in_int:
                     mask[i] = 1
                 else:
                     if convert_to_masked_nullable:
                         mask[i] = 1
-                    seen.float_ = True
+                seen.saw_null(seen_float=not allow_null_in_int)
             else:
-                seen.float_ = True
-            floats[i] = complexes[i] = fval
+                seen.set_curr_val(fval)
         elif util.is_integer_object(val):
-            floats[i] = complexes[i] = val
-
             val = int(val)
             seen.saw_int(val)
 
             if val >= 0:
                 if val <= oUINT64_MAX:
-                    uints[i] = val
+                    seen.set_curr_val(<uint64_t>val)
```

Comment:
Are these casts necessary? I thought we got rid of them before - maybe a mistake in merge that brought them back? Generally should not have a need for casting - only weakens the type system further

Reply:
Thanks for the review.
Casting should be required here. 
val is a Python object, so Cython wouldn't know how to implicitly cast here.
(We also do special things like casting bool -> uint8, and the logic to decide to cast to int64/uint64 you're referring to here).